spark/BidResponseJob.scala

```diff
@@ -0,0 +1,191 @@
+package com.liveintent.dwh.logs.jobs
+
+import com.liveintent.dwh.dmp.models._
+import com.liveintent.dwh.dmp.sources.BidderGroupInfoSource
+import com.liveintent.dwh.spark.util.{SparkArgs, SparkDwhJob}
+import com.liveintent.dwh.userver.sources.ImpressionSource
+import com.liveintent.util.logparser.LogParser
+import com.twitter.algebird.Moments
+import org.apache.spark.rdd.RDD
+
+
+class BidResponsesJob(override val args: SparkArgs) extends SparkDwhJob(args) {
+
+  import BidResponsesJob._
+
+  private lazy val impressions = ImpressionSource()
+    .map{ _.sanitizeToPositive }
+    .filter { i => i.isFirstImpression && i.demandType == Some("default") }
+    .groupBy { e => (e.decisionId.getOrElse(""), e.redecisionAttempt.getOrElse(-1)) }
+    .mapValues(_.take(1))
+    .map( _._2 )
+
+  private lazy val brokenDropTriplets = impressions.filter( _.isBrokenDrop )
+    .map( i => (Seq(i.md5,i.sh1,i.sh2).flatMap( e => e ).filter( e => e != null && !e.isEmpty ).headOption.getOrElse(""),
+      i.templateId.getOrElse(-1),
+      i.placementId) )
+    .distinct
+    .collect()
+    .toSet
+
+  private lazy val purchaseGroups = impressions
+    .map(i => {
+      val keys = (
+        Seq(i.md5, i.sh1, i.sh2).flatMap(e => e).filter(e => e != null && !e.isEmpty).headOption.getOrElse(""),
+        i.templateId.getOrElse(-1),
+        i.placementId)
+      if (brokenDropTriplets.contains(keys)) i.copy(isBrokenDrop = true) else i
+    })
+    .addPrefixGroup({ e => (e.bidderId, e.emailhash.map( _.value )) })
+    .getGroupAggregates(
+      momentsColumns= {
+        case i => Map(
+          "spend" -> i.advertiserSpent.getOrElse(0.0),
+          "clearingPrice" -> i.clearingPrice.getOrElse(0.0),
+          "expProfit" -> Seq(Some(0.0), i.actualDSPFee, i.actualPGMFee, i.actualSSPFee, i.thirdpartyMargin).flatMap(e => e).sum
+        )
+      },
+      sumColumns={ case i => if (i.bidderId.isDefined) Map("isRtb"->1) else Map() },
+      countColumns= {
+        case i => Map(
+          "pricingTypes" -> i.pricingType.filterNot(_.isEmpty).getOrElse("UNKNOWN-PRICINGTYPE"),
+          "deviceTypes" -> i.userAgent.map(LogParser.parseUa(_).getDeviceCategory().getName()).filterNot(dt => dt.isEmpty || dt.toLowerCase() == "unknown").getOrElse("UNKNOWN-DEVICETYPE"),
+          "oss" -> i.userAgent.map(LogParser.parseUa(_).getOperatingSystem().getFamilyName()).filterNot(os => os.isEmpty || os.toLowerCase() == "unknown").getOrElse("UNKNOWN-OS"),
+          "browsers" -> i.userAgent.map(LogParser.parseUa(_).getFamily().getName()).filterNot(b => b.isEmpty || b.toLowerCase() == "unknown").getOrElse("UNKNOWN-BROWSER"),
+          "sourcesImp" -> i.buyeruidSource.filterNot(_.isEmpty).getOrElse("UNKNOWN-UIDSOURCE")
+        )
+      }
+    )
+
+  private lazy val processingDate = jobDates.lastDay.toStrings().head
```

Comment:
Remember to ask the candidate why we do this (in case they do not bring it up themselves)